include/TPP/Dialect/Perf/PerfOps.td

```diff
@@ -0,0 +1,261 @@
+// BenchOp
+//===----------------------------------------------------------------------===//
+def Perf_BenchOp : Perf_Op<"bench",
+    [AutomaticAllocationScope, SingleBlockImplicitTerminator<"perf::YieldOp">]> {
+  let summary = "Benchmark the enclosed code.";
+  let description = [{
+    The `perf.bench` operation generates benchmarking code
+    around the enclosed region.
+    The performance results are gathered over the specified
+    number of iterations.
+
+    The operation allocates memory to store the results.
+    The lifetime of the returned results has to be managed by the user.
+
+    Example:
+
+    ```mlir
+
+    %deltas = perf.bench (%n) {
+      ... // ops under measurement
+    } -> memref<?xf64>
+
+    ```
+
+    Optionally, a yield op can be added:
+
+    ```mlir
+
+    %deltas = perf.bench (%n) {
+      ... // ops under measurement
+      perf.yield
+    } -> memref<?xf64>
+
+    ```
+  }];
+
+  let arguments = (ins I64:$numIters);
+  let results = (outs RankedOrUnrankedMemRefOf<[F64]>:$deltas,
+            Variadic<AnyType>:$bodyResults);
+  let regions = (region SizedRegion<1>:$region);
+
+  let assemblyFormat = [{
+    `(` $numIters `)`
+    $region attr-dict
+    `->` type($deltas) (`,` type($bodyResults)^)?
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// YieldOp
+//===----------------------------------------------------------------------===//
+
+def Perf_YieldOp : Perf_Op<"yield", [ HasParent<"BenchOp">,
+              Pure, Terminator, ReturnLike]> {
+  let summary = "Yield values to parent operation.";
+  let description = [{
+    The `perf.yield` operation yields an SSA value from the perf dialect op
+    region and terminates the regions. The semantics of how the values are
+    yielded is defined by the parent operation.
+    If `perf.yield` has any operands, the operands must match the parent
+    operation's results.
+    If the parent operation defines no values, then the `perf.yield` may be
+    left out in the custom syntax and the builders will insert one implicitly.
+    Otherwise, it has to be present in the syntax to indicate which values are
+    yielded.
+  }];
+
+  let arguments = (ins Variadic<AnyType>:$results);
+  let builders = [OpBuilder<(ins), [{ /* nothing to do */ }]>];
+
+  let assemblyFormat =
+    [{  attr-dict ($results^ `:` type($results))? }];
+
+  let hasVerifier = 1;
+}
+
+//===----------------------------------------------------------------------===//
+// StartTimerOp
+//===----------------------------------------------------------------------===//
+
+def Perf_StartTimerOp : Perf_Op<"start_timer", []> {
+  let summary = "Start a timer.";
+  let description = [{
+    The `perf.start_timer` operation creates a new timer
+    which begins time measurement.
+
+    See `perf.stop_timer` for timer termination.
+
+    Example:
+
+    ```mlir
+
+    %timer = perf.start_timer : i64
```

Comment:
What is the semantics of this `i64`? Unique? auto-increment? tied to some implementation detail?

Reply:
Right now it models unique timer instances in line with the current benchmark timer design.
As discussed, I moved to a custom `!perf.timer` type. For lowering to function calls, it will probably still default to `i64` that will represent a unique timer ID.